File: ios/Nerox/WireGuardTunnel.m
#import "WireGuardTunnel.h"
#import <NetworkExtension/NetworkExtension.h>

static NSError *WireGuardBridgeError(NSString *message, NSError *underlyingError)
{
  NSMutableDictionary *userInfo = [NSMutableDictionary dictionaryWithObject:message ?: @"WireGuard tunnel failed"
                                                                     forKey:NSLocalizedDescriptionKey];
  if (underlyingError != nil) {
    userInfo[NSUnderlyingErrorKey] = underlyingError;
  }

  return [NSError errorWithDomain:@"WireGuardTunnel"
                             code:underlyingError != nil ? underlyingError.code : 1
                         userInfo:userInfo];
}

@implementation WireGuardTunnel

RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(start,
                 startWithConfig:(NSString *)config
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  if (config.length == 0) {
    reject(@"WG_CONFIG_EMPTY", @"WireGuard config is empty", nil);
    return;
  }

  [NETunnelProviderManager loadAllFromPreferencesWithCompletionHandler:^(NSArray<NETunnelProviderManager *> *managers, NSError *error) {
    if (error != nil) {
      NSString *message = error.localizedDescription ?: @"Failed to load iOS VPN configurations";
      reject(@"WG_LOAD_FAILED", message, WireGuardBridgeError(message, error));
      return;
    }

    NETunnelProviderManager *manager = managers.firstObject ?: [[NETunnelProviderManager alloc] init];
    NETunnelProviderProtocol *protocol = [[NETunnelProviderProtocol alloc] init];
    NSString *bundleId = [[NSBundle mainBundle] bundleIdentifier];
    protocol.providerBundleIdentifier = [NSString stringWithFormat:@"%@.PacketTunnel", bundleId];
    protocol.serverAddress = @"Nerox WireGuard";
    protocol.providerConfiguration = @{
      @"wgQuickConfig": config,
      @"tunnelName": @"nerox"
    };

    manager.protocolConfiguration = protocol;
    manager.localizedDescription = @"Nerox VPN";
    manager.enabled = YES;

    [manager saveToPreferencesWithCompletionHandler:^(NSError *saveError) {
      if (saveError != nil) {
        NSString *message = [NSString stringWithFormat:@"%@. Check that the main app is signed with the Network Extension entitlement.",
                             saveError.localizedDescription ?: @"Failed to save iOS VPN configuration"];
        reject(@"WG_SAVE_FAILED", message, WireGuardBridgeError(message, saveError));
        return;
      }

      [manager loadFromPreferencesWithCompletionHandler:^(NSError *reloadError) {
        if (reloadError != nil) {
          NSString *message = reloadError.localizedDescription ?: @"Failed to reload iOS VPN configuration";
          reject(@"WG_RELOAD_FAILED", message, WireGuardBridgeError(message, reloadError));
          return;
        }

        NSError *startError = nil;
        BOOL started = [manager.connection startVPNTunnelAndReturnError:&startError];
        if (!started || startError != nil) {
          NSString *message = [NSString stringWithFormat:@"%@. Check that the PacketTunnel extension target is embedded and signed, and that its bundle id matches %@.PacketTunnel.",
                               startError.localizedDescription ?: @"Failed to start iOS VPN tunnel",
                               bundleId];
          reject(@"WG_START_FAILED", message, WireGuardBridgeError(message, startError));
          return;
        }

        resolve(@"UP");
      }];
    }];
  }];
}

RCT_REMAP_METHOD(stop,
                 stopWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  [NETunnelProviderManager loadAllFromPreferencesWithCompletionHandler:^(NSArray<NETunnelProviderManager *> *managers, NSError *error) {
    if (error != nil) {
      NSString *message = error.localizedDescription ?: @"Failed to load iOS VPN configurations";
      reject(@"WG_LOAD_FAILED", message, WireGuardBridgeError(message, error));
      return;
    }

    for (NETunnelProviderManager *manager in managers) {
      [manager.connection stopVPNTunnel];
    }

    resolve(@YES);
  }];
}

RCT_REMAP_METHOD(getStatus,
                 getStatusWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  [NETunnelProviderManager loadAllFromPreferencesWithCompletionHandler:^(NSArray<NETunnelProviderManager *> *managers, NSError *error) {
    if (error != nil) {
      NSString *message = error.localizedDescription ?: @"Failed to load iOS VPN configurations";
      reject(@"WG_LOAD_FAILED", message, WireGuardBridgeError(message, error));
      return;
    }

    NETunnelProviderManager *manager = managers.firstObject;
    if (manager == nil) {
      resolve(@"DISCONNECTED");
      return;
    }

    switch (manager.connection.status) {
      case NEVPNStatusConnected:
        resolve(@"UP");
        break;
      case NEVPNStatusConnecting:
        resolve(@"CONNECTING");
        break;
      case NEVPNStatusDisconnecting:
        resolve(@"DISCONNECTING");
        break;
      default:
        resolve(@"DOWN");
        break;
    }
  }];
}

RCT_REMAP_METHOD(getStatistics,
                 getStatisticsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  [NETunnelProviderManager loadAllFromPreferencesWithCompletionHandler:^(NSArray<NETunnelProviderManager *> *managers, NSError *error) {
    if (error != nil || managers.count == 0) {
      resolve(@{@"totalReceived": @(0), @"totalSent": @(0)});
      return;
    }

    NETunnelProviderManager *manager = managers.firstObject;
    NETunnelProviderSession *session = (NETunnelProviderSession *)manager.connection;

    if (session == nil || session.status != NEVPNStatusConnected) {
      resolve(@{@"totalReceived": @(0), @"totalSent": @(0)});
      return;
    }

    // Send a message to the PacketTunnel extension requesting traffic stats
    NSError *sendError = nil;
    NSData *messageData = [@"getTransferredByteCount" dataUsingEncoding:NSUTF8StringEncoding];
    [session sendProviderMessage:messageData
                    returnError:&sendError
                responseHandler:^(NSData * _Nullable responseData) {
      if (responseData == nil) {
        // Extension didn't respond with stats - return zeros
        resolve(@{@"totalReceived": @(0), @"totalSent": @(0)});
        return;
      }

      NSError *jsonError = nil;
      NSDictionary *stats = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:&jsonError];
      if (jsonError != nil || stats == nil) {
        resolve(@{@"totalReceived": @(0), @"totalSent": @(0)});
        return;
      }

      NSNumber *rx = stats[@"totalReceived"] ?: @(0);
      NSNumber *tx = stats[@"totalSent"] ?: @(0);
      resolve(@{@"totalReceived": rx, @"totalSent": tx});
    }];

    if (sendError != nil) {
      // If sendProviderMessage fails, return zeros gracefully
      resolve(@{@"totalReceived": @(0), @"totalSent": @(0)});
    }
  }];
}

@end

File: ios/WireGuardTunnel.swift
import Foundation
import NetworkExtension

@objc(WireGuardTunnel)
class WireGuardTunnel: NSObject {

  private var manager: NETunnelProviderManager?

  @objc
  static func requiresMainQueueSetup() -> Bool { return false }

  // JS se: WireGuardTunnel.start(config)
  @objc(start:resolver:rejecter:)
  func start(_ config: String,
             resolver resolve: @escaping RCTPromiseResolveBlock,
             rejecter reject: @escaping RCTPromiseRejectBlock) {

    NETunnelProviderManager.loadAllFromPreferences { managers, error in
      if let error = error {
        reject("WG_LOAD_FAILED", error.localizedDescription, error)
        return
      }

      let mgr = managers?.first ?? NETunnelProviderManager()

      let proto = NETunnelProviderProtocol()
      proto.providerBundleIdentifier = "org.reactjs.native.example.Nerox.PacketTunnel"
      proto.serverAddress = "Nerox VPN"
      proto.providerConfiguration = [
        "wgQuickConfig": config,
        "tunnelName": "nerox"
      ]

      mgr.protocolConfiguration = proto
      mgr.localizedDescription = "Nerox VPN"
      mgr.isEnabled = true

      // YEH LINE pehli baar VPN permission popup laati hai 👇
      mgr.saveToPreferences { error in
        if let error = error {
          reject("WG_SAVE_FAILED", error.localizedDescription, error)
          return
        }
        // Save ke baad load karna zaroori hai
        mgr.loadFromPreferences { error in
          if let error = error {
            reject("WG_RELOAD_FAILED", error.localizedDescription, error)
            return
          }
          do {
            try mgr.connection.startVPNTunnel()
            self.manager = mgr
            resolve("STARTED")
          } catch {
            reject("WG_START_FAILED", error.localizedDescription, error)
          }
        }
      }
    }
  }

  @objc(stop:rejecter:)
  func stop(_ resolve: @escaping RCTPromiseResolveBlock,
            rejecter reject: @escaping RCTPromiseRejectBlock) {
    NETunnelProviderManager.loadAllFromPreferences { managers, error in
      managers?.first?.connection.stopVPNTunnel()
      resolve(true)
    }
  }

  @objc(getStatus:rejecter:)
  func getStatus(_ resolve: @escaping RCTPromiseResolveBlock,
                 rejecter reject: @escaping RCTPromiseRejectBlock) {
    NETunnelProviderManager.loadAllFromPreferences { managers, error in
      guard let mgr = managers?.first else { resolve("DISCONNECTED"); return }
      switch mgr.connection.status {
      case .connected:    resolve("CONNECTED")
      case .connecting:   resolve("CONNECTING")
      case .reasserting:  resolve("CONNECTING")
      default:            resolve("DISCONNECTED")
      }
    }
  }

  @objc(getStatistics:rejecter:)
  func getStatistics(_ resolve: @escaping RCTPromiseResolveBlock,
                     rejecter reject: @escaping RCTPromiseRejectBlock) {
    NETunnelProviderManager.loadAllFromPreferences { managers, error in
      guard let session = managers?.first?.connection as? NETunnelProviderSession else {
        resolve(["totalReceived": 0, "totalSent": 0]); return
      }
      do {
        try session.sendProviderMessage("getTransferredByteCount".data(using: .utf8)!) { data in
          if let data = data,
             let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            resolve(json)
          } else {
            resolve(["totalReceived": 0, "totalSent": 0])
          }
        }
      } catch {
        resolve(["totalReceived": 0, "totalSent": 0])
      }
    }
  }
}